core/trino-main/src/main/java/io/trino/sql/planner/PlanFragmenter.java

```diff
@@ -508,13 +559,28 @@ else if (exchange.getType() == ExchangeNode.Type.REPARTITION) {
                     isWorkerCoordinatorBoundary(context.get(), childrenProperties.build()) ? getRetryPolicy(session) : RetryPolicy.NONE);
         }
 
-        private SubPlan buildSubPlan(PlanNode node, FragmentProperties properties, RewriteContext<FragmentProperties> context)
+        private SubPlan buildSubPlan(PlanNode node, FragmentIdentifier fragmentIdentifier, FragmentProperties properties, RewriteContext<FragmentProperties> context)
```

Comment:
This dihotomy of `FragmentIdentifier` and `PlanFragmentId` is confusing.

Reply:
Also why do we need to rebuild fragment for "unchanged" ones. Would it be more logical to cache whole unchanged `SubPlan` and return here instead of rebuilding?